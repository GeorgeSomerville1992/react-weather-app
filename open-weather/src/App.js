import React, { Component } from 'react';
import { apiEndpoint } from './utils/api';
import WeatherContent from './components/weather-content/index';
import helper from './utils/helpers';
import WeatherNavigation from './components/weather-navigation/index';
import { Container, Header, Content, Footer } from './styles';

class App extends Component {

  state = {
    isFetching: false,
    error: null,
    chosenCity: 'London',
    weatherContent: helper.getNextFiveDays(),
  }

  componentDidMount() {
    this.getWeather();
  }

  componentWillUpdate(nextProps, nextState) {
    if(nextState.chosenCity !== this.state.chosenCity) {
      this.getWeather(nextState.chosenCity)
    }
  }

  handleCityChange = (city) => () => {
    this.setState(() => ({
      chosenCity: city
    }));
  }

  handleSelectApiDataSource = () => {
    this.getWeather('London')
  }

  handleGetWeatherFromCsvFile = csvFile => {
    this.setState(() => ({
      isFetching: false,
      weatherContent: helper.filterData(helper.getNextFiveDays(), csvFile)
    }));
  }

  getWeather() {
    this.setState({
      isFetching: true,
    }, () => {
      apiEndpoint(this.state.chosenCity).getWeather.then(response => {
        this.setState(() => ({
          isFetching: false,
          weatherContent: helper.filterData(helper.getNextFiveDays(), response.data)
        }));
      }).catch(error => {
        this.setState(prevState => ({
          isFetching: false,
          error: error,
        }));
      }).then(() => {
        this.setState({
          isFetching: false,
        })
      });
    });
  }

  render() {
    const { isFetching, chosenCity, weatherContent } = this.state

    return (
      <Container>
        <Header>
          { isFetching ? 'Loading....' : chosenCity }
        </Header>
        <Content>
          { !isFetching && <WeatherContent days={weatherContent} />}
        </Content>
        <Footer>
          <WeatherNavigation
            handleCityChange={this.handleCityChange}
            handleGetWeatherFromCsvFile={this.handleGetWeatherFromCsvFile}
            handleSelectApiDataSource={this.handleSelectApiDataSource}
            />
        </Footer>
      </Container>
    );
  }
}

export default App;
