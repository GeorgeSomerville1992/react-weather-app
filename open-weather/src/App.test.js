import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { shallow, mount, configure } from 'enzyme';
import App from './App';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

jest.mock('./utils/api', () => {
  return {
    apiEndpoint: jest.fn(() => {
      return {
        getWeather: Promise.resolve({
          data: {
            list: [
              {
                dt: 1531872000,
                main: {
                  temp : 293.05
                }
              },
              {
                dt: 1531882800,
                main: {
                  temp : 293.05
                }
              },
              {
                dt: 1531893600,
                main: {
                  temp : 293.05
                }
              },
            ]
          }
        }),
      }
    })
  }
})

jest.mock('./utils/helpers', () => {
  return {
    getNextFiveDays: jest.fn(() => {
      return [
        {
          currentDay: 'Monday',
          times: []
        },
        {
          currentDay: 'Tuesday',
          times: [
              {
                dt: 1531785600,
                main: {
                  temp: 293.05,
                },
              },
              {
                dt: 1531796400,
                main: {
                  temp: 293.05,
                },
              },
            ],
        },
        {
          currentDay: 'Wednesday',
          times: []
        },
        {
          currentDay: 'Thursday',
          times: []
        },
        {
          currentDay: 'Friday',
          times: []
        }
      ]
    }),
    convertTodegreesCelsius: jest.fn(() => {
      return 33
    })
  }
})


describe('Weather app tests', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });


  it('renders loading the table without mock data', async () => {
    const app = await mount(<App />);
    await app.instance().componentDidMount();
    expect(app.debug()).toMatchSnapshot();
  })

  it('renders with correct initial state', () => {
    const wrapper = shallow(<App/>);
    const status = wrapper.state();

    expect(wrapper.state().isFetching).toBe(true);
    expect(wrapper.state().error).toBe(null);
    expect(wrapper.state().chosenCity).toBe('London');
    expect(wrapper.state().weatherContent).toHaveLength(5);
  })

  it('handles the getWeather function correctly on mount', () => {
    const getWeatherSpy = jest.fn();
    const wrapper = shallow(<App />);
    wrapper.instance().getWeather = getWeatherSpy;
    wrapper.instance().componentDidMount();
    expect(getWeatherSpy).toHaveBeenCalled();
  })

  it('handles get weather request from api correctly and filters the data', async () => {
    const wrapper = shallow(<App />)
    await wrapper.instance().componentDidMount();
    expect(wrapper.state().weatherContent[1].times).toHaveLength(2);
  })

  it('correctly changes the country to get weather upon function call', async () => {
    const wrapper = shallow(<App />)
    await wrapper.instance().componentDidMount();
    expect(wrapper.state().chosenCity).toBe('London');
    wrapper.instance().handleCityChange('Liverpool')();
    expect(wrapper.state().chosenCity).toBe('Liverpool');
  })
})