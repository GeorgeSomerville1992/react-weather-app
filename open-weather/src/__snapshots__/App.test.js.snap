// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Weather app tests renders loading the table without mock data 1`] = `
"<App>
  <styled.div>
    <div className=\\"sc-gzVnrw fiZJyz\\">
      <styled.div>
        <div className=\\"sc-htoDjs bxSFQy\\">
          Loading....
        </div>
      </styled.div>
      <styled.div>
        <div className=\\"sc-dnqmqq euuIvu\\" />
      </styled.div>
      <styled.div>
        <div className=\\"sc-iwsKbI czDOXk\\">
          <WeatherNavigation handleCityChange={[Function]} handleGetWeatherFromCsvFile={[Function]} handleSelectApiDataSource={[Function]}>
            <styled.div>
              <div className=\\"sc-bZQynM gZCdoJ\\">
                <styled.div>
                  <div className=\\"sc-EHOje eItBph\\">
                     Select a City 
                  </div>
                </styled.div>
                <styled.div>
                  <div className=\\"sc-ifAKCX ZmwtQ\\">
                    <WithStyles(Button) onClick={[Function]}>
                      <Button classes={{...}} onClick={[Function]} color=\\"default\\" component=\\"button\\" disabled={false} disableFocusRipple={false} fullWidth={false} mini={false} size=\\"medium\\" type=\\"button\\" variant=\\"text\\">
                        <WithStyles(ButtonBase) className=\\"MuiButton-root-145 MuiButton-text-147 MuiButton-flat-150\\" disabled={false} focusRipple={true} focusVisibleClassName=\\"MuiButton-focusVisible-162\\" onClick={[Function]} component=\\"button\\" type=\\"button\\">
                          <ButtonBase classes={{...}} className=\\"MuiButton-root-145 MuiButton-text-147 MuiButton-flat-150\\" disabled={false} focusRipple={true} focusVisibleClassName=\\"MuiButton-focusVisible-162\\" onClick={[Function]} component=\\"button\\" type=\\"button\\" centerRipple={false} disableRipple={false} disableTouchRipple={false} tabIndex=\\"0\\">
                            <button onBlur={[Function: handleEvent]} onFocus={[Function]} onKeyDown={[Function]} onKeyUp={[Function]} onMouseDown={[Function: handleEvent]} onMouseLeave={[Function: handleEvent]} onMouseUp={[Function: handleEvent]} onTouchEnd={[Function: handleEvent]} onTouchMove={[Function: handleEvent]} onTouchStart={[Function: handleEvent]} tabIndex=\\"0\\" className=\\"MuiButtonBase-root-169 MuiButton-root-145 MuiButton-text-147 MuiButton-flat-150\\" type=\\"button\\" disabled={false} onClick={[Function]}>
                              <span className=\\"MuiButton-label-146\\">
                                London
                              </span>
                              <WithStyles(TouchRipple) innerRef={[Function]} center={false}>
                                <TouchRipple classes={{...}} center={false}>
                                  <TransitionGroup component=\\"span\\" enter={true} exit={true} className=\\"MuiTouchRipple-root-172\\" childFactory={[Function: childFactory]}>
                                    <span className=\\"MuiTouchRipple-root-172\\" />
                                  </TransitionGroup>
                                </TouchRipple>
                              </WithStyles(TouchRipple)>
                            </button>
                          </ButtonBase>
                        </WithStyles(ButtonBase)>
                      </Button>
                    </WithStyles(Button)>
                    <WithStyles(Button) onClick={[Function]}>
                      <Button classes={{...}} onClick={[Function]} color=\\"default\\" component=\\"button\\" disabled={false} disableFocusRipple={false} fullWidth={false} mini={false} size=\\"medium\\" type=\\"button\\" variant=\\"text\\">
                        <WithStyles(ButtonBase) className=\\"MuiButton-root-145 MuiButton-text-147 MuiButton-flat-150\\" disabled={false} focusRipple={true} focusVisibleClassName=\\"MuiButton-focusVisible-162\\" onClick={[Function]} component=\\"button\\" type=\\"button\\">
                          <ButtonBase classes={{...}} className=\\"MuiButton-root-145 MuiButton-text-147 MuiButton-flat-150\\" disabled={false} focusRipple={true} focusVisibleClassName=\\"MuiButton-focusVisible-162\\" onClick={[Function]} component=\\"button\\" type=\\"button\\" centerRipple={false} disableRipple={false} disableTouchRipple={false} tabIndex=\\"0\\">
                            <button onBlur={[Function: handleEvent]} onFocus={[Function]} onKeyDown={[Function]} onKeyUp={[Function]} onMouseDown={[Function: handleEvent]} onMouseLeave={[Function: handleEvent]} onMouseUp={[Function: handleEvent]} onTouchEnd={[Function: handleEvent]} onTouchMove={[Function: handleEvent]} onTouchStart={[Function: handleEvent]} tabIndex=\\"0\\" className=\\"MuiButtonBase-root-169 MuiButton-root-145 MuiButton-text-147 MuiButton-flat-150\\" type=\\"button\\" disabled={false} onClick={[Function]}>
                              <span className=\\"MuiButton-label-146\\">
                                Liverpool
                              </span>
                              <WithStyles(TouchRipple) innerRef={[Function]} center={false}>
                                <TouchRipple classes={{...}} center={false}>
                                  <TransitionGroup component=\\"span\\" enter={true} exit={true} className=\\"MuiTouchRipple-root-172\\" childFactory={[Function: childFactory]}>
                                    <span className=\\"MuiTouchRipple-root-172\\" />
                                  </TransitionGroup>
                                </TouchRipple>
                              </WithStyles(TouchRipple)>
                            </button>
                          </ButtonBase>
                        </WithStyles(ButtonBase)>
                      </Button>
                    </WithStyles(Button)>
                    <WithStyles(Button) onClick={[Function]}>
                      <Button classes={{...}} onClick={[Function]} color=\\"default\\" component=\\"button\\" disabled={false} disableFocusRipple={false} fullWidth={false} mini={false} size=\\"medium\\" type=\\"button\\" variant=\\"text\\">
                        <WithStyles(ButtonBase) className=\\"MuiButton-root-145 MuiButton-text-147 MuiButton-flat-150\\" disabled={false} focusRipple={true} focusVisibleClassName=\\"MuiButton-focusVisible-162\\" onClick={[Function]} component=\\"button\\" type=\\"button\\">
                          <ButtonBase classes={{...}} className=\\"MuiButton-root-145 MuiButton-text-147 MuiButton-flat-150\\" disabled={false} focusRipple={true} focusVisibleClassName=\\"MuiButton-focusVisible-162\\" onClick={[Function]} component=\\"button\\" type=\\"button\\" centerRipple={false} disableRipple={false} disableTouchRipple={false} tabIndex=\\"0\\">
                            <button onBlur={[Function: handleEvent]} onFocus={[Function]} onKeyDown={[Function]} onKeyUp={[Function]} onMouseDown={[Function: handleEvent]} onMouseLeave={[Function: handleEvent]} onMouseUp={[Function: handleEvent]} onTouchEnd={[Function: handleEvent]} onTouchMove={[Function: handleEvent]} onTouchStart={[Function: handleEvent]} tabIndex=\\"0\\" className=\\"MuiButtonBase-root-169 MuiButton-root-145 MuiButton-text-147 MuiButton-flat-150\\" type=\\"button\\" disabled={false} onClick={[Function]}>
                              <span className=\\"MuiButton-label-146\\">
                                Manchester
                              </span>
                              <WithStyles(TouchRipple) innerRef={[Function]} center={false}>
                                <TouchRipple classes={{...}} center={false}>
                                  <TransitionGroup component=\\"span\\" enter={true} exit={true} className=\\"MuiTouchRipple-root-172\\" childFactory={[Function: childFactory]}>
                                    <span className=\\"MuiTouchRipple-root-172\\" />
                                  </TransitionGroup>
                                </TouchRipple>
                              </WithStyles(TouchRipple)>
                            </button>
                          </ButtonBase>
                        </WithStyles(ButtonBase)>
                      </Button>
                    </WithStyles(Button)>
                    <WithStyles(Button) onClick={[Function]}>
                      <Button classes={{...}} onClick={[Function]} color=\\"default\\" component=\\"button\\" disabled={false} disableFocusRipple={false} fullWidth={false} mini={false} size=\\"medium\\" type=\\"button\\" variant=\\"text\\">
                        <WithStyles(ButtonBase) className=\\"MuiButton-root-145 MuiButton-text-147 MuiButton-flat-150\\" disabled={false} focusRipple={true} focusVisibleClassName=\\"MuiButton-focusVisible-162\\" onClick={[Function]} component=\\"button\\" type=\\"button\\">
                          <ButtonBase classes={{...}} className=\\"MuiButton-root-145 MuiButton-text-147 MuiButton-flat-150\\" disabled={false} focusRipple={true} focusVisibleClassName=\\"MuiButton-focusVisible-162\\" onClick={[Function]} component=\\"button\\" type=\\"button\\" centerRipple={false} disableRipple={false} disableTouchRipple={false} tabIndex=\\"0\\">
                            <button onBlur={[Function: handleEvent]} onFocus={[Function]} onKeyDown={[Function]} onKeyUp={[Function]} onMouseDown={[Function: handleEvent]} onMouseLeave={[Function: handleEvent]} onMouseUp={[Function: handleEvent]} onTouchEnd={[Function: handleEvent]} onTouchMove={[Function: handleEvent]} onTouchStart={[Function: handleEvent]} tabIndex=\\"0\\" className=\\"MuiButtonBase-root-169 MuiButton-root-145 MuiButton-text-147 MuiButton-flat-150\\" type=\\"button\\" disabled={false} onClick={[Function]}>
                              <span className=\\"MuiButton-label-146\\">
                                Oxford
                              </span>
                              <WithStyles(TouchRipple) innerRef={[Function]} center={false}>
                                <TouchRipple classes={{...}} center={false}>
                                  <TransitionGroup component=\\"span\\" enter={true} exit={true} className=\\"MuiTouchRipple-root-172\\" childFactory={[Function: childFactory]}>
                                    <span className=\\"MuiTouchRipple-root-172\\" />
                                  </TransitionGroup>
                                </TouchRipple>
                              </WithStyles(TouchRipple)>
                            </button>
                          </ButtonBase>
                        </WithStyles(ButtonBase)>
                      </Button>
                    </WithStyles(Button)>
                    <WithStyles(Button) onClick={[Function]}>
                      <Button classes={{...}} onClick={[Function]} color=\\"default\\" component=\\"button\\" disabled={false} disableFocusRipple={false} fullWidth={false} mini={false} size=\\"medium\\" type=\\"button\\" variant=\\"text\\">
                        <WithStyles(ButtonBase) className=\\"MuiButton-root-145 MuiButton-text-147 MuiButton-flat-150\\" disabled={false} focusRipple={true} focusVisibleClassName=\\"MuiButton-focusVisible-162\\" onClick={[Function]} component=\\"button\\" type=\\"button\\">
                          <ButtonBase classes={{...}} className=\\"MuiButton-root-145 MuiButton-text-147 MuiButton-flat-150\\" disabled={false} focusRipple={true} focusVisibleClassName=\\"MuiButton-focusVisible-162\\" onClick={[Function]} component=\\"button\\" type=\\"button\\" centerRipple={false} disableRipple={false} disableTouchRipple={false} tabIndex=\\"0\\">
                            <button onBlur={[Function: handleEvent]} onFocus={[Function]} onKeyDown={[Function]} onKeyUp={[Function]} onMouseDown={[Function: handleEvent]} onMouseLeave={[Function: handleEvent]} onMouseUp={[Function: handleEvent]} onTouchEnd={[Function: handleEvent]} onTouchMove={[Function: handleEvent]} onTouchStart={[Function: handleEvent]} tabIndex=\\"0\\" className=\\"MuiButtonBase-root-169 MuiButton-root-145 MuiButton-text-147 MuiButton-flat-150\\" type=\\"button\\" disabled={false} onClick={[Function]}>
                              <span className=\\"MuiButton-label-146\\">
                                Newcastle
                              </span>
                              <WithStyles(TouchRipple) innerRef={[Function]} center={false}>
                                <TouchRipple classes={{...}} center={false}>
                                  <TransitionGroup component=\\"span\\" enter={true} exit={true} className=\\"MuiTouchRipple-root-172\\" childFactory={[Function: childFactory]}>
                                    <span className=\\"MuiTouchRipple-root-172\\" />
                                  </TransitionGroup>
                                </TouchRipple>
                              </WithStyles(TouchRipple)>
                            </button>
                          </ButtonBase>
                        </WithStyles(ButtonBase)>
                      </Button>
                    </WithStyles(Button)>
                  </div>
                </styled.div>
                <styled.div>
                  <div className=\\"sc-EHOje eItBph\\">
                     Select a Data Source 
                  </div>
                </styled.div>
                <styled.div>
                  <div className=\\"sc-ifAKCX ZmwtQ\\">
                    <a label=\\"csv\\" onFileLoaded={[Function]}>
                      <div className=\\"csv-reader-input\\">
                        <label>
                          csv
                        </label>
                        <input className=\\"csv-input\\" type=\\"file\\" id={{...}} accept=\\"text/csv\\" onChange={[Function: onChange]} />
                      </div>
                    </a>
                    <WithStyles(Button) onClick={[Function]}>
                      <Button classes={{...}} onClick={[Function]} color=\\"default\\" component=\\"button\\" disabled={false} disableFocusRipple={false} fullWidth={false} mini={false} size=\\"medium\\" type=\\"button\\" variant=\\"text\\">
                        <WithStyles(ButtonBase) className=\\"MuiButton-root-145 MuiButton-text-147 MuiButton-flat-150\\" disabled={false} focusRipple={true} focusVisibleClassName=\\"MuiButton-focusVisible-162\\" onClick={[Function]} component=\\"button\\" type=\\"button\\">
                          <ButtonBase classes={{...}} className=\\"MuiButton-root-145 MuiButton-text-147 MuiButton-flat-150\\" disabled={false} focusRipple={true} focusVisibleClassName=\\"MuiButton-focusVisible-162\\" onClick={[Function]} component=\\"button\\" type=\\"button\\" centerRipple={false} disableRipple={false} disableTouchRipple={false} tabIndex=\\"0\\">
                            <button onBlur={[Function: handleEvent]} onFocus={[Function]} onKeyDown={[Function]} onKeyUp={[Function]} onMouseDown={[Function: handleEvent]} onMouseLeave={[Function: handleEvent]} onMouseUp={[Function: handleEvent]} onTouchEnd={[Function: handleEvent]} onTouchMove={[Function: handleEvent]} onTouchStart={[Function: handleEvent]} tabIndex=\\"0\\" className=\\"MuiButtonBase-root-169 MuiButton-root-145 MuiButton-text-147 MuiButton-flat-150\\" type=\\"button\\" disabled={false} onClick={[Function]}>
                              <span className=\\"MuiButton-label-146\\">
                                API
                              </span>
                              <WithStyles(TouchRipple) innerRef={[Function]} center={false}>
                                <TouchRipple classes={{...}} center={false}>
                                  <TransitionGroup component=\\"span\\" enter={true} exit={true} className=\\"MuiTouchRipple-root-172\\" childFactory={[Function: childFactory]}>
                                    <span className=\\"MuiTouchRipple-root-172\\" />
                                  </TransitionGroup>
                                </TouchRipple>
                              </WithStyles(TouchRipple)>
                            </button>
                          </ButtonBase>
                        </WithStyles(ButtonBase)>
                      </Button>
                    </WithStyles(Button)>
                  </div>
                </styled.div>
              </div>
            </styled.div>
          </WeatherNavigation>
        </div>
      </styled.div>
    </div>
  </styled.div>
</App>"
`;
